### PR TITLE
add support for Spring Framework 6 AOT engine 

### PR DESCRIPTION
**Pull request by joshlong**
Hi 

Please find a proof-of-concept introduction of AOT support in Spring Boot 3, per #2398 .

I've tested the proposed changes and they work with both Spring Boot 3 and Spring Boot 2. 

Spring Boot 3 is not yet GA, however. So, i've had to add `snapshot` and `milestone` repositories. Obviously, this won't do in a library like the Kubernetes Java client which needs to be releasable all the time. I'm not sure how you want to handle this. Maybe we create a separate module that doesn't get released and that has the appropriate repositories? Maybe we just stash this PR until [November 24th](https://github.com/spring-projects/spring-boot/milestone/240), when Spring Boot 3 is due to go GA? (maybe a little later than that, who knows...) 
 
Also, for my contribution to work, I need to add a dependency on an external [library called Reflections](https://www.baeldung.com/reflections-library). This is used to capture all the code-generated types for CRDs, for example, in the Kubernetes API _and_ in the user's auto-configured packages in a Spring Boot application. This library is _only_ used at compile time, and _only_ when the user is using Spring Boot 3 AOT compilation. that's not the default. I don't know how to handle that, either. I suppose you could make the package `optional` and then make sure that the interested developer who wnats to use AOT processing gets a warning or something when they try it without that class on the classpath? I don't like that approach very much. It'd be nice if the library was just there and it worked for anybody who wants to build an AOT/ GraalVM native image, out-of-the-box.  

Is there a Maven `scope` i don't know about that limits the lifetime of the binary _explicitly_ to the time the AOT engine runs? 

I tried this new support out on [a sample Kubernetes CRD+Controller demo](https://github.com/kubernetes-native-java/controllers-101) and it works just fine, better than what I was able to do before with Spring Native, actually. From ac consumer's perspective, I didn't have to do anything to make code from the Kubernetes Java client do its magic. I was able to delete code, even. 

thanks for your time and consideration

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -482,49 +487,49 @@
               <include>src/test/groovy/**/*.groovy</include>
             </includes>
 
-            <importOrder>  <!-- or a custom ordering -->
```
Comment (brendandburns) [inline pom.xml:485]: Please remove these whitespace changes.

Comment (brendandburns) [inline pom.xml:485]: Please remove these whitespace changes.

File: spring/src/main/java/io/kubernetes/client/spring/extended/controller/config/aot/KubernetesBeanFactoryInitializationAotProcessor.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+    private final MemberCategory[] allMemberCategories = MemberCategory.values();
+
+    @Override
+    public BeanFactoryInitializationAotContribution processAheadOfTime(
+            @NotNull ConfigurableListableBeanFactory beanFactory) {
+        return (generationContext, beanFactoryInitializationCode) -> {
+            RuntimeHints hints = generationContext.getRuntimeHints();
+            String[] classNames = new String[]{
+                    "com.google.gson.JsonElement",//
```
Comment (brendandburns) [inline spring/src/main/java/io/kubernetes/client/spring/extended/controller/config/aot/KubernetesBeanFactoryInitializationAotProcessor.java:38]: Is there a reason why there are empty comments after each of these lines?

**Comment by brendandburns** [comment]
I took a look at the code changes and other than the whitespace changes which need to be reverted, this looks fine to me.

I will let @yue9944882 take a look at the Spring side and comment on how we want to sequence this into the released client.

File: pom.xml (hunk 1 above)
Comment (joshlong) [reply pom.xml]: i believe this has now been addressed.

Comment (joshlong) [reply pom.xml]: i believe this has now been addressed.

**Comment by joshlong** [comment]
Hi, thanks for the quick replies. somehow, the biggest change of my original PR - the fact that it needs a snapshot or milestone (non-Maven central) repository was not in the original `pom.xml`. Don't ask me how. So I took the code on `main`, applied it over mine, _re_-applied the changes without accidentally invoking my IDE's code reformatter, and here we are. 

I've updated the PR

**Comment by brendandburns** [comment]
We already have a spring component (spring-boot-actuator) marked as `optional` https://github.com/kubernetes-client/java/blob/master/pom.xml#L177 so I think that for the reflections dependency, it makes sense mark it optional also, although I don't feel super strongly about this, but we've definitely gotten feedback about not wanting this library to pull in lots and lots of dependencies.

Regarding the milestone dependency, you're definitely right that we don't want to pull this in until it is released. So it feels like we should leave this PR open until it is released. There's not too much overlap w/ this PR so it shouldn't be too hard to keep it rebased.

**Comment by joshlong** [comment]
Awesome! Both of those points - delaying until Boot 3 GA and marking Reflections as `optional` - are  reasonable and I appreciate them. I'll follow up sometime after Spring Boot 3 is GA. 

Maybe we could do something where, if you try to run the code and do an AOT compilation, it can check that reflections is on the classpath and if it isn't then warn the user to add it to their classpath? That shouldn't be too hard. Spring has a `boolean ClassUtils.isPresent (String className, ClassLoader)` we can put the warning behind. We'd have to update that one class in my PR. 

I'm pretty pleased how far this has come in just a week. Thanks for a great project, and I appreciate the fast turnarounds. Hopefully, people interested in doing so will soon (before 2023, maybe?) be able to build Kubernetes clients and controllers that take 50mb of RAM and startup In no time at all soon enough.

**Comment by brendandburns** [comment]
Adding that warning seems totally fine to me.

Thanks

**Comment by yue9944882** [comment]
> Error:  COMPILATION ERROR : 
Error:  /home/runner/work/java/java/spring/src/main/java/io/kubernetes/client/spring/extended/controller/config/KubernetesInformerProperties.java:[17,51] cannot access org.springframework.boot.context.properties.ConfigurationProperties
  bad class file: /home/runner/.m2/repository/org/springframework/boot/spring-boot/3.0.0-M5/spring-boot-3.0.0-M5.jar(org/springframework/boot/context/properties/ConfigurationProperties.class)
    class file has wrong version 61.0, should be 52.0
    
@joshlong the idea SGTM, but i noticed the error messages above in the build log. is spring-boot 3 still compatible with java 8? there's still a distinct portion of users actively using java 8, so the compatibility should be preserved as well.

**Comment by joshlong** [comment]
> > Error:  COMPILATION ERROR : 
> 
> Error:  /home/runner/work/java/java/spring/src/main/java/io/kubernetes/client/spring/extended/controller/config/KubernetesInformerProperties.java:[17,51] cannot access org.springframework.boot.context.properties.ConfigurationProperties
> 
>   bad class file: /home/runner/.m2/repository/org/springframework/boot/spring-boot/3.0.0-M5/spring-boot-3.0.0-M5.jar(org/springframework/boot/context/properties/ConfigurationProperties.class)
> 
>     class file has wrong version 61.0, should be 52.0
> 
>     
> 
> @joshlong the idea SGTM, but i noticed the error messages above in the build log. is spring-boot 3 still compatible with java 8? there's still a distinct portion of users actively using java 8, so the compatibility should be preserved as well.

Hi - no, spring boot 3 requires a Java 17 baseline.

**Comment by brendandburns** [comment]
I filed an issue so we can discuss how we want to handle Spring 6 and it's dependency on Java 17

https://github.com/kubernetes-client/java/issues/2452

**Comment by yue9944882** [comment]
/close

closing it b/c #2457 was merged

**Comment by OlgaMaciaszek** [comment]
Hi @brendandburns, when is a release containing these changes scheduled for?

**Pull request closed without merging**